# Toxic Comment Classification

As a NLP Engineer, I've always been interested in working on maulicious comments such as cyberbullying, abuse, hate speech, violent threats to create safe and healthy cyperspace and to protect kids from tragic outcomes.  

In this work, I built a binary classifcation model which is capable of detecting toxic comments. 
The data are downloaded from toxic comment classificaiton challenge dataset from Kaggle which contains omments from Wikipedia’s talk page edits. The training data contains 159571 and test contains 153164 comments and they are classified as either toxic (class = 1) or non toxic (class = 0).

I used this training dataset to train a model that can take in new comments in test dataset and predict whether they are toxic or not.
The data also include different types of toxicity which are toxic, severe toxic, threats, obscenity, insults, and identity-based hate. I would like to further investigate and build multiclassificaiton model as a next step in the near future. 

In [1]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer 
from sklearn.naive_bayes import MultinomialNB

df_train = pd.read_csv('train.csv')
df_train.head(15)

id                                       comment_text  \
0   0000997932d777bf  Explanation\nWhy the edits made under my usern...   
1   000103f0d9cfb60f  D'aww! He matches this background colour I'm s...   
2   000113f07ec002fd  Hey man, I'm really not trying to edit war. It...   
3   0001b41b1c6bb37e  "\nMore\nI can't make any real suggestions on ...   
4   0001d958c54c6e35  You, sir, are my hero. Any chance you remember...   
5   00025465d4725e87  "\n\nCongratulations from me as well, use the ...   
6   0002bcb3da6cb337       COCKSUCKER BEFORE YOU PISS AROUND ON MY WORK   
7   00031b1e95af7921  Your vandalism to the Matt Shirvington article...   
8   00037261f536c51d  Sorry if the word 'nonsense' was offensive to ...   
9   00040093b2687caa  alignment on this subject and which are contra...   
10  0005300084f90edc  "\nFair use rationale for Image:Wonju.jpg\n\nT...   
11  00054a5e18b50dd4  bbq \n\nbe a man and lets discuss it-maybe ove...   
12  0005c987bdfc9d4b  Hey... what is it..\n@ | talk .\nWhat is it......   
13  0006f16e4e9f292e  Before you start throwing accusations and warn...   
14  00070ef96486d6f9  Oh, and the girl above started her arguments w...   

    toxic  severe_toxic  obscene  threat  insult  identity_hate  
0       0             0        0       0       0              0  
1       0             0        0       0       0              0  
2       0             0        0       0       0              0  
3       0             0        0       0       0              0  
4       0             0        0       0       0              0  
5       0             0        0       0       0              0  
6       1             1        1       0       1              0  
7       0             0        0       0       0              0  
8       0             0        0       0       0              0  
9       0             0        0       0       0              0  
10      0             0        0       0       0              0  
11      0             0        0       0       0              0  
12      1             0        0       0       0              0  
13      0             0        0       0       0              0  
14      0             0        0       0       0              0

I was curisous how bad the comments look like when all the categories are classified as 1 so here it is.

In [2]:
bad_bad_comment = 'toxic == 1 & severe_toxic ==1 & obscene == 1 & threat == 1 & insult == 1 & identity_hate == 1' 
bad_comments = df_train.query(bad_bad_comment)
bad_comments

id                                       comment_text  \
1017    02c6e41e4b317ac3  WOULDN'T BE THE FIRST TIME BITCH. FUCK YOU I'L...   
1312    039296aa294ee10b  SHUT UP, YOU FAT POOP, OR I WILL KICK YOUR ASS!!!   
7299    1368c10281978876  You're a stupid cunt \n\nFuck you dumb arse, y...   
13648   2405c66d017f139e  Bitch \n\nYou are a little bitch. I fuckin spe...   
13964   24d2b50726b67167  I am going to murder ZimZalaBim ST47 for being...   
22158   3a704e2efa9fa300  FUCK YOU!!!!!!!!!!!! YOU FUCKING NIGGER BAG OF...   
29968   4f95394c7229f84c  u motherfukkin bitch i want to rape you smelly...   
32098   554a87fb67e51344  Fuck All Asyriac Nation \n\nQamishli belong to...   
33951   5a8eab9cfe601d3a  GO FUCK YOURSELF BITCH.  I HATE YOUR SOULD.  M...   
38513   66c3bced1735cb90  AM GOING TO RAPE YOU IN THE ASS YOU FAT BITCH ...   
38578   66f0a9006c188820  fuck you honkey, why you hatin' on blacks? You...   
46155   7b50a142ad5a428d        So fuck off and die, rape fantasising sluts   
55160   93569e0a8ce776b8  ANYONE WHO SUPPORTS THIS IS FUCKING SICK. MEN ...   
59279   9ecf3329158e0276  JEW \n\nGet the fuck out of here you jewish so...   
65075   ae1b1f693e4913b5  FAGGOTS!  YO FUCKER IT WAS FUCKING HAYES YOU F...   
67247   b3e7b8a6ea14ddc5  YOURE A FRREAKING JEW, AND PLEAZE COMMIT SUICIDE.   
73821   c586b7a2fd575b13  Shut up you asswipe, we don't care. I'll decap...   
77139   ce9204582479520c  Eat shit you fucking arse rapping jew fucking ...   
80612   d7aaabfd8799b464  I hit a jack off all day, after the rape my mo...   
82141   dbbbb42bc096d8dc  Help! \n\nIf you become an admin, I will fucki...   
99819   161bd9408c74abb7  Glenn Beck is a faggot. He likes dick in his a...   
100222  188ec3b7baf66495  Urgent!!!! \n\nMind yo own fucken buisiness st...   
106617  3a4c7758fad18de3  , I hope your retarded kids get anal raped and...   
116767  7032e0dce02a180f  Go fuck yourself!!! you fucking delteted it an...   
118928  7ba7b175e597c13e  You god damn son of a bitch you fucking blocke...   
123794  9622c82f97eb26cd  She is a Motherfucker \n\nI will block you Mot...   
133551  ca7c493a2a81b65f  and your little faggot boy Propol, pray, pray ...   
142513  fa3c0dc48563a1f8  faggot \n\nYou lil piece of shit I haven’t van...   
151131  77d84b1321c22d9a  LGBT \n\nyou little fuck , are you a fag , tha...   
157428  ddf6dc5cf6931f48  Ok.... \n\nBitch i swear to God i will fuckin ...   
159400  fd052883fa6a8697  Shalom \n\nSemite, get the fuck out of here. I...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
1017        1             1        1       1       1              1  
1312        1             1        1       1       1              1  
7299        1             1        1       1       1              1  
13648       1             1        1       1       1              1  
13964       1             1        1       1       1              1  
22158       1             1        1       1       1              1  
29968       1             1        1       1       1              1  
32098       1             1        1       1       1              1  
33951       1             1        1       1       1              1  
38513       1             1        1       1       1              1  
38578       1             1        1       1       1              1  
46155       1             1        1       1       1              1  
55160       1             1        1       1       1              1  
59279       1             1        1       1       1              1  
65075       1             1        1       1       1              1  
67247       1             1        1       1       1              1  
73821       1             1        1       1       1              1  
77139       1             1        1       1       1              1  
80612       1             1        1       1       1              1  
82141       1             1        1       1       1              1  
99819       1          

Also how big the proportion of toxic comments in a whole dataset? 

In [3]:
toxic = df_train[df_train['toxic'] == 1]
len(toxic)/len(df_train) 

0.09584448302009764

9.6% of comments are toxic! It's almost 10% of data :( 

And I also wanted to see the size in the differnt combination of toxicity. 

In [4]:
count = df_train.groupby(['toxic', 'severe_toxic', 'obscene', 'threat', 'insult', 'identity_hate']).size() 
count

toxic  severe_toxic  obscene  threat  insult  identity_hate
0      0             0        0       0       0                143346
                                              1                    54
                                      1       0                   301
                                              1                    28
                              1       0       0                    22
                                      1       0                     3
                     1        0       0       0                   317
                                              1                     3
                                      1       0                   181
                                              1                    18
                              1       0       0                     2
                                      1       0                     2
1      0             0        0       0       0                  5666
                              

First step is to create a word vector using BOW (bag of words) with stopwords filter.  

In [5]:
word_bagger = CountVectorizer(stop_words="english")
word_bagger.fit(df_train['comment_text'])

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words='english',
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [6]:
total_words = word_bagger.get_feature_names()
len(total_words)

189460

For this work, I decided to use Multinomial Naive Bayes model as it is suitable for data where the features are counts. 

In [7]:
y = df_train['toxic']
X = word_bagger.transform(df_train['comment_text'])
toxic_nb = MultinomialNB()
toxic_nb.fit(X, y)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [8]:
toxic_nb.score(X,y)

0.9551798259082164

95.5% training accuracy is predicted.
Now apply test data to the trained model. 

In [9]:
df_test = pd.read_csv('test.csv')
df_test.head()

id                                       comment_text
0  00001cee341fdb12  Yo bitch Ja Rule is more succesful then you'll...
1  0000247867823ef7  == From RfC == \n\n The title is fine as it is...
2  00013b17ad220c46  " \n\n == Sources == \n\n * Zawe Ashton on Lap...
3  00017563c3f7919a  :If you have a look back at the source, the in...
4  00017695ad8997eb          I don't anonymously edit articles at all.

In [10]:
test_bagger = word_bagger.transform(df_test['comment_text'])
test_bagger.shape

(153164, 189460)

In [11]:
df_test['predicted'] = toxic_nb.predict(test_bagger)
df_test['probability'] = toxic_nb.predict_proba(test_bagger)[:,1]
df_test

id                                       comment_text  \
0       00001cee341fdb12  Yo bitch Ja Rule is more succesful then you'll...   
1       0000247867823ef7  == From RfC == \n\n The title is fine as it is...   
2       00013b17ad220c46  " \n\n == Sources == \n\n * Zawe Ashton on Lap...   
3       00017563c3f7919a  :If you have a look back at the source, the in...   
4       00017695ad8997eb          I don't anonymously edit articles at all.   
...                  ...                                                ...   
153159  fffcd0960ee309b5  . \n i totally agree, this stuff is nothing bu...   
153160  fffd7a9a6eb32c16  == Throw from out field to home plate. == \n\n...   
153161  fffda9e8d6fafa9e  " \n\n == Okinotorishima categories == \n\n I ...   
153162  fffe8f1340a79fc2  " \n\n == ""One of the founding nations of the...   
153163  ffffce3fb183ee80  " \n :::Stop already. Your bullshit is not wel...   

        predicted   probability  
0               1  1.000000e+00  
1               0  3.921071e-04  
2               0  3.242127e-02  
3               0  4.254467e-09  
4               0  2.636425e-03  
...           ...           ...  
153159          0  6.980053e-02  
153160          0  9.874606e-06  
153161          0  6.431109e-12  
153162          0  2.665747e-16  
153163          1  9.743731e-01  

[153164 rows x 4 columns]

Let's see if it works with my own constructed data.

In [12]:
c1 = "I think coronavirus makes people nuts."
c2 = "ppl are fucking selfish pieces of shit."
c3 = "my dog is a bitch." # true statement
c4 = "japanese ppl don't shower." # sarcasm! 
c5 = "I think you are very ugly." # why not toxic? 
test_bagger2 = word_bagger.transform([c1,c2,c3,c4,c5])

In [13]:
toxic_nb.predict(test_bagger2)

array([0, 1, 1, 1, 0])

In [14]:
toxic_nb.predict_proba(test_bagger2)

array([[9.00060764e-01, 9.99392356e-02],
       [2.22451906e-05, 9.99977755e-01],
       [6.91199941e-03, 9.93088001e-01],
       [4.74873887e-01, 5.25126113e-01],
       [6.40072471e-01, 3.59927529e-01]])

In [15]:
toxic_nb.predict_proba(test_bagger2)[:,1]

array([0.09993924, 0.99997775, 0.993088  , 0.52512611, 0.35992753])

It looks pretty good overall but there's definitely some room for improvement. For instance, c5 can be read very ambigious and c4 is a sarcasm and not necessarily toxic which is known as one of the hardest chalelnges NLP appliaction is dealing with.

### Evaluation

Toxic comment classification challenge data from Kaggle also contains a dataset for evaluation conveniently so I compared my model's prediction against the evalution dataset.

In [16]:
df_test_label = pd.read_csv('test_labels.csv')
df_test_label

id  toxic  severe_toxic  obscene  threat  insult  \
0       00001cee341fdb12     -1            -1       -1      -1      -1   
1       0000247867823ef7     -1            -1       -1      -1      -1   
2       00013b17ad220c46     -1            -1       -1      -1      -1   
3       00017563c3f7919a     -1            -1       -1      -1      -1   
4       00017695ad8997eb     -1            -1       -1      -1      -1   
...                  ...    ...           ...      ...     ...     ...   
153159  fffcd0960ee309b5     -1            -1       -1      -1      -1   
153160  fffd7a9a6eb32c16     -1            -1       -1      -1      -1   
153161  fffda9e8d6fafa9e     -1            -1       -1      -1      -1   
153162  fffe8f1340a79fc2     -1            -1       -1      -1      -1   
153163  ffffce3fb183ee80     -1            -1       -1      -1      -1   

        identity_hate  
0                  -1  
1                  -1  
2                  -1  
3                  -1  
4                  -1  
...               ...  
153159             -1  
153160             -1  
153161             -1  
153162             -1  
153163             -1  

[153164 rows x 7 columns]

-1 means these comments are not used for evaluation so I only extracted data where 'toxic' value is 0 (=non toxic) or 1 (=toxic) and added a column 'matched' to compare my model prediciton and manually evaluated result.  

In [17]:
df_evaluated = df_test_label[df_test_label['toxic'] > -1]
df_final = pd.merge(df_evaluated, df_test, how='left', on='id')
df_final['matched'] = np.where(df_final['toxic'] == df_final['predicted'], 'True', 'False')  
df_final

id  toxic  severe_toxic  obscene  threat  insult  \
0      0001ea8717f6de06      0             0        0       0       0   
1      000247e83dcc1211      0             0        0       0       0   
2      0002f87b16116a7f      0             0        0       0       0   
3      0003e1cccfd5a40a      0             0        0       0       0   
4      00059ace3e3e9a53      0             0        0       0       0   
...                 ...    ...           ...      ...     ...     ...   
63973  fff8f64043129fa2      0             0        0       0       0   
63974  fff9d70fe0722906      0             0        0       0       0   
63975  fffa8a11c4378854      0             0        0       0       0   
63976  fffac2a094c8e0e2      1             0        1       0       1   
63977  fffb5451268fb5ba      0             0        0       0       0   

       identity_hate                                       comment_text  \
0                  0  Thank you for understanding. I think very high...   
1                  0                   :Dear god this site is horrible.   
2                  0  "::: Somebody will invariably try to add Relig...   
3                  0  " \n\n It says it right there that it IS a typ...   
4                  0  " \n\n == Before adding a new product to the l...   
...              ...                                                ...   
63973              0  :Jerome, I see you never got around to this…! ...   
63974              0  ==Lucky bastard== \n http://wikimediafoundatio...   
63975              0  ==shame on you all!!!== \n\n You want to speak...   
63976              0  MEL GIBSON IS A NAZI BITCH WHO MAKES SHITTY MO...   
63977              0  " \n\n == Unicorn lair discovery == \n\n Suppo...   

       predicted   probability matched  
0              0  7.566100e-05    True  
1              0  4.127167e-01    True  
2              0  5.662520e-14    True  
3              0  6.457884e-19    True  
4              0  3.152840e-16    True  
...          ...           ...     ...  
63973          0  1.393345e-10    True  
63974          1  9.973388e-01   False  
63975          0  3.435685e-01    True  
63976          1  9.999993e-01    True  
63977          0  9.646441e-03    True  

[63978 rows x 11 columns]

In [18]:
matched = df_final.matched.value_counts().loc['True']
matched/len(df_final)

0.9180343242989778

Overall accuracy is 91.8%! 
To summarize, I present performance measures of confusion matrix, precision, recall and F1 score. 

In [19]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score 
from sklearn.metrics import classification_report 

actual = df_final['toxic']
predicted = df_final['predicted']
results = confusion_matrix(actual,predicted)
TP = results[1,1]
TN = results[0,0]
FP = results[0,1]
FN = results[1,0]
print ('Confusion Matrix :')
print(results) 
print('[[TN ','FP','\n  FN',' TP]]')
print ('\nAccuracy Score :',accuracy_score(actual, predicted)) 
print ('\nReport : ')
print (classification_report(actual, predicted)) 

Confusion Matrix :
[[54656  3232]
 [ 2012  4078]]
[[TN  FP 
  FN  TP]]

Accuracy Score : 0.9180343242989778

Report : 
              precision    recall  f1-score   support

           0       0.96      0.94      0.95     57888
           1       0.56      0.67      0.61      6090

    accuracy                           0.92     63978
   macro avg       0.76      0.81      0.78     63978
weighted avg       0.93      0.92      0.92     63978



<b>Classification Rate/Accuracy:</b><br>
Accuracy = (TP + TN) / (TP + TN + FP + FN)<br>
         = (4078 + 54656) /(4078 + 54656 + 3232 + 2012)<br> 
         = 0.9180343242989778

<b>Recall</b>: <i>Recall gives us an idea about when it’s actually yes, how often does it predict yes.</i><br>
Recall = TP / (TP + FN)<br>
       = 4078 / (4078 + 2012)<br> 
       = 0.6696223316912973

<b>Precision</b>: <i>Precsion tells us about when it predicts yes, how often is it correct.</i><br>
Precision = TP / (TP + FP)<br>
          = 4078 / (4078 + 3232)<br>
          = 0.5578659370725034

<b>F1-score:</b> <i>The weighted harmonic mean of the test's precision and recall</i><br>
F1-score = (2 * Recall * Precision) / (Recall + Presision)<br>
         = (2 * 0.67 * 0.56) / (0.67 + 0.56)<br> 
         = 0.6100813008130082